File: src/main/java/com/camackenzie/exvi/core/util/CachedKey.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.camackenzie.exvi.core.util;

import java.security.Key;
import javax.crypto.spec.SecretKeySpec;

/**
 *
 * @author callum
 */
public class CachedKey {

    private final String algorithm, encoded;

    public CachedKey(Key key) {
        this.algorithm = CryptographyUtils.bytesToBase64String(key.getAlgorithm().getBytes());
        this.encoded = CryptographyUtils.bytesToBase64String(key.getEncoded());
    }

    public Key getKey() {
        String algo = new String(CryptographyUtils.bytesFromBase64String(this.algorithm));
        byte[] enc = CryptographyUtils.bytesFromBase64String(this.encoded);
        return new SecretKeySpec(enc, algo);
    }

}

File: exvi-core/src/commonMain/kotlin/com/camackenzie/exvi/core/util/CachedKey.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.camackenzie.exvi.core.util

import com.soywiz.krypto.SecureRandom
import kotlinx.serialization.*
import kotlinx.serialization.json.*
import com.soywiz.krypto.encoding.Base64
import com.soywiz.krypto.encoding.toBase64
import kotlin.jvm.JvmStatic

/**
 *
 * @author callum
 */
@Serializable
data class CachedKey(val key: String) {
    fun getKeyAsBytes(): ByteArray = Base64.decode(key)

    val bytes
        get() = getKeyAsBytes()

    companion object {
        @JvmStatic
        fun generateKey(): CachedKey = generateKey(32)

        @JvmStatic
        fun generateKey(nBytes: Int) = CachedKey(SecureRandom.nextBytes(nBytes).toBase64())
    }
}